### PR TITLE
Support default toJSON in WebIDL

### LINKED ISSUE #22781: Support default toJSON in WebIDL
https://searchfox.org/mozilla-central/rev/c07aaf12f13037c1f5a343d31f8291549e57373f/dom/bindings/Codegen.py#8846-8879
https://searchfox.org/mozilla-central/rev/c07aaf12f13037c1f5a343d31f8291549e57373f/dom/bindings/Configuration.py#422-441

### PR DESCRIPTION
**Pull request by saschanaz**
Ported related lines from gecko.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #22781

- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23844)

### CODE REVIEW

**Comment by saschanaz** [comment]
Mind having a look? 😁

**Comment by Manishearth** [comment]
Reviewing now, sorry for the delay, had some more urgent work items I needed to deal with before rustconf/vacation.

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 1]
```diff
@@ -2881,6 +2881,49 @@ def __str__(self):
...
+                Argument('*const %s' % descriptor.concreteType, 'this'),
+                Argument('&RootedGuard<*mut JSObject>', 'result')]
+        CGAbstractMethod.__init__(self, descriptor, 'CollectJSONAttributes',
+                                  'bool', args, pub=True, unsafe=True)
+        self.toJSONMethod = toJSONMethod
+
+    def definition_body(self):
+        ret = ''
+        interface = self.descriptor.interface
+        for m in interface.members:
```
Comment (Manishearth) [inline components/script/dom/bindings/codegen/CodegenRust.py]: does this deal with inheritance? Should we make it deal with inheritance?

**Review by Manishearth — APPROVED** [review]
Code looks good, I'm unsure about inheritance. r=me if you think it doesn't need to deal with it

File: components/script/dom/bindings/codegen/CodegenRust.py (hunk 1 above)
Comment (saschanaz) [reply components/script/dom/bindings/codegen/CodegenRust.py]: Inheritance is supported; not here but at https://github.com/servo/servo/pull/23844/files#diff-60d01595cff328c165842fea9e4ccbc2R3684. Each interface defines its `CollectJSONAttributes` which is called by ToJSON in an ordered way.

**Comment by Manishearth** [comment]
@bors-servo r+

thanks!

**Comment by saschanaz** [comment]
Hmm, I don't think the fetch test failure is relevant... 🤔

**Comment by CYBAI** [comment]
@bors-servo retry

- it's #23290

**Pull request merged**